test/plugin/base_test.rb

```diff
@@ -1391,6 +1412,7 @@ def test_k8s_monitored_resource_fallback
         expected_params: K8S_CONTAINER_PARAMS
       },
       # When local_resource_id is not present or does not match k8s regexes.
+      # TODO these two are busted.
```

Comment:
For option 3:
`k8s_container_log_entry` mocks the actual log entry format Kubernetes generates. My concern is that if we add another artificial field, we are no longer testing what we should be testing.

For option 1:
Agreed that it's not the right direction.

For option 2:
We just need to change line https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/blob/dcf2e0f6bd6cf9a35a754af13d963729aac789e7/test/plugin/base_test.rb#L1482 to verify against `textPayload` if the log entry fall backs to use `gke_container`. This might not be that expensive.

Reply:
Hmm... In fact, I just verified that this change below will make the test pass. But I don't think this is the expected behavior.

```
  # Need to add this param.
  CONTAINER_FROM_FALLBACK_PARAMS = {
    resource: {
      type: GKE_CONSTANTS[:resource_type],
      labels: {
        # When falling back to 'gke_container' resource type, we are missing
        # a few required labels: 'namespace_id', 'pod_id', 'container_name' because the legacy
        # path rely on `tag` to follow certain regex:
        # '\.(?<pod_name>[^_]+)_(?<namespace_name>[^_]+)_(?<container_name>.+)$'
        'cluster_name' => CONTAINER_CLUSTER_NAME,
        'instance_id' => VM_ID,
        'zone' => ZONE
      }
    },
    log_name: 'test',
    project_id: PROJECT_ID,
    labels: {
      "#{GKE_CONSTANTS[:service]}/stream" => CONTAINER_STREAM,
      "#{COMPUTE_CONSTANTS[:service]}/resource_name" => HOSTNAME
    }
  }.freeze
...
      {
        config: ENABLE_METADATA_AGENT_CONFIG,
        setup_metadata_agent_stub: true,
        setup_k8s_stub: true,
        log_entry: k8s_container_log_entry(
          log_entry(0)).reject { |k, _| k == LOCAL_RESOURCE_ID_KEY },
        expected_params: CONTAINER_FROM_FALLBACK_PARAMS
      },
      {
        config: ENABLE_METADATA_AGENT_CONFIG,
        setup_metadata_agent_stub: true,
        setup_k8s_stub: true,
        log_entry: k8s_container_log_entry(
          log_entry(0),
          local_resource_id: RANDOM_LOCAL_RESOURCE_ID),
        # When 'kube-env' is present, "compute.googleapis.com/resource_name" is
        # not added.
        expected_params: CONTAINER_FROM_FALLBACK_PARAMS
      },
...
      if test_params[:expected_params] == CONTAINER_FROM_FALLBACK_PARAMS
        verify_log_entries(1, test_params[:expected_params],
                           'textPayload') do |entry|
          field = entry['textPayload']
          assert_equal 'test log entry 0', field, entry
        end
      end
```

Reply:
Ok, this below should work:
```
      {
        config: ENABLE_METADATA_AGENT_CONFIG,
        setup_metadata_agent_stub: true,
        setup_k8s_stub: true,
        log_entry: k8s_container_log_entry(
          log_entry(0)).reject { |k, _| k == LOCAL_RESOURCE_ID_KEY },
        expected_params: CONTAINER_FROM_TAG_PARAMS
      },
      {
        config: ENABLE_METADATA_AGENT_CONFIG,
        setup_metadata_agent_stub: true,
        setup_k8s_stub: true,
        log_entry: k8s_container_log_entry(
          log_entry(0),
          local_resource_id: RANDOM_LOCAL_RESOURCE_ID),
        # When 'kube-env' is present, "compute.googleapis.com/resource_name" is
        # not added.
        expected_params: CONTAINER_FROM_TAG_PARAMS
      },
...
      # used to miss the CONTAINER_TAG.
      d = create_driver(test_params[:config], CONTAINER_TAG)
...
      verify_log_entries(1, test_params[:expected_params],
                           'textPayload') do |entry|
        field = entry['textPayload']
        assert_equal 'test log entry 0', field, entry
      end
```

To fix the test, we need to add `CONTAINER_TAG` for `k8s_container` tests.

I just submit a PR https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/252 to split the tests into `k8s_container` and `k8s_node` and only add `CONTAINER_TAG` for `k8s_container`. Please rebase once that is merged.

Reply:
https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/229 has been merged.